## Grouping Data & Aggregating Variables

### Step 1. Import the data sets into Jupyter

### Step 2. Grouping Data with Pandas

### Step 3. Create a Loyalty Flag

### Step 4. Analyze Spending Habits by Loyalty Category


### Step 5. Create a Spending Flag


### Step 6. Create an Order Frequency Flag

### Step 7. Export the Dataframe in pickle Format to the "Prepared Data" Folder

## Import libraries

In [7]:
# Import libraries

import pandas as pd
import numpy as np
import os

## Step 1. Import the data sets into Jupyter

In [8]:
# Tell Python to remember a main folder path
path = r'/Users/ashwanisherawat/Downloads/Instacart Basket Analysis/02 Data'

In [9]:
path

'/Users/ashwanisherawat/Downloads/Instacart Basket Analysis/02 Data'

In [16]:
# Import and identify dataframes
ords_prods_merge = pd.read_pickle(os.path.join(path, 'Prepared Data', 'orders_products_merged_derived.pkl'))

In [17]:
ords_prods_merge.shape

(32435059, 17)

In [20]:
# Create a subset of 1 million rows
df = ords_prods_merge[:1000000]

In [21]:
# Check shape of new subset
df.shape

(1000000, 17)

In [22]:
df.head(10)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   
5   2398795        1             2                   3                  7   
6   2398795        1             2                   3                  7   
7   2398795        1             2                   3                  7   
8   2398795        1             2                   3                  7   
9   2398795        1             2                   3                  7   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                     NaN         196                  1          0   both   
1                     NaN       14084                  2          0   both   
2                     NaN       12427                  3          0   both   
3                     NaN       26088                  4          0   both   
4                     NaN       26405                  5          0   both   
5                    15.0         196                  1          1   both   
6                    15.0       10258                  2          0   both   
7                    15.0       12427                  3          1   both   
8                    15.0       13176                  4          0   both   
9                    15.0       26088                  5          1   both   

   Unnamed: 0                             product_name  aisle_id  \
0       195.0                                     Soda      77.0   
1     14084.0  Organic Unsweetened Vanilla Almond Milk      91.0   
2     12427.0                      Original Beef Jerky      23.0   
3     26089.0               Aged White Cheddar Popcorn      23.0   
4     26406.0         XL Pick-A-Size Paper Towel Rolls      54.0   
5       195.0                                     Soda      77.0   
6     10258.0                               Pistachios     117.0   
7     12427.0                      Original Beef Jerky      23.0   
8     13176.0                   Bag of Organic Bananas      24.0   
9     26089.0               Aged White Cheddar Popcorn      23.0   

   department_id  prices     busiest_day busiest_period_of_day  
0            7.0     9.0  Regularly busy        Average orders  
1           16.0    12.5  Regularly busy        Average orders  
2           19.0     4.4  Regularly busy        Average orders  
3           19.0     4.7  Regularly busy        Average orders  
4           17.0     1.0  Regularly busy        Average orders  
5            7.0     9.0  Regularly busy        Average orders  
6           19.0     3.0  Regularly busy        Average orders  
7           19.0     4.4  Regularly busy        Average orders  
8            4.0    10.3  Regularly busy        Average orders  
9           19.0     4.7  Regularly busy        Average orders

## Step 2. Grouping Data with Pandas

In [23]:
# The 'groupby()' function will group a given dataframe by a given column
df.groupby('product_name')

In [24]:
df.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1.0              14.792915
2.0              17.091743
3.0              17.919589
4.0              17.891973
5.0              15.214270
6.0              15.382228
7.0              17.699986
8.0              16.485269
9.0              15.965921
10.0             20.091818
11.0             16.485395
12.0             15.615845
13.0             16.485052
14.0             17.523474
15.0             15.690354
16.0             18.014342
17.0             16.155822
18.0             19.606536
19.0             17.630236
20.0             17.138204
21.0             21.996844

In [25]:
df.groupby('department_id')['order_number'].mean()


department_id
1.0     14.792915
2.0     17.091743
3.0     17.919589
4.0     17.891973
5.0     15.214270
6.0     15.382228
7.0     17.699986
8.0     16.485269
9.0     15.965921
10.0    20.091818
11.0    16.485395
12.0    15.615845
13.0    16.485052
14.0    17.523474
15.0    15.690354
16.0    18.014342
17.0    16.155822
18.0    19.606536
19.0    17.630236
20.0    17.138204
21.0    21.996844
Name: order_number, dtype: float64

## Comment:

Used both mehtod mentioned in the learning content and the analysis shows department 21 has the highest average orders (21.99), followed by departments 10 (20.09) and 18 (19.61), indicating high demand. Department 1 has the lowest average (14.79), suggesting less frequent purchases. This highlights varying demand levels across departments for targeted planning.

In [26]:
df.groupby('department_id').agg({'order_number': ['mean', 'min', 'max']})

order_number        
                      mean min max
department_id                     
1.0              14.792915   1  99
2.0              17.091743   1  98
3.0              17.919589   1  99
4.0              17.891973   1  99
5.0              15.214270   1  99
6.0              15.382228   1  99
7.0              17.699986   1  99
8.0              16.485269   1  91
9.0              15.965921   1  99
10.0             20.091818   1  99
11.0             16.485395   1  99
12.0             15.615845   1  99
13.0             16.485052   1  99
14.0             17.523474   1  99
15.0             15.690354   1  99
16.0             18.014342   1  99
17.0             16.155822   1  99
18.0             19.606536   1  99
19.0             17.630236   1  99
20.0             17.138204   1  99
21.0             21.996844   1  97

## Step 3. Create a Loyalty Flag

In [33]:
# Aggregating Data with transform() Function
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

In [34]:
ords_prods_merge.head(15)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2539329        1             1                   2                  8   
2    2539329        1             1                   2                  8   
3    2539329        1             1                   2                  8   
4    2539329        1             1                   2                  8   
5    2398795        1             2                   3                  7   
6    2398795        1             2                   3                  7   
7    2398795        1             2                   3                  7   
8    2398795        1             2                   3                  7   
9    2398795        1             2                   3                  7   
10   2398795        1             2                   3                  7   
11    473747        1             3                   3                 12   
12    473747        1             3                   3                 12   
13    473747        1             3                   3                 12   
14    473747        1             3                   3                 12   

    days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                      NaN         196                  1          0   both   
1                      NaN       14084                  2          0   both   
2                      NaN       12427                  3          0   both   
3                      NaN       26088                  4          0   both   
4                      NaN       26405                  5          0   both   
5                     15.0         196                  1          1   both   
6                     15.0       10258                  2          0   both   
7                     15.0       12427                  3          1   both   
8                     15.0       13176                  4          0   both   
9                     15.0       26088                  5          1   both   
10                    15.0       13032                  6          0   both   
11                    21.0         196                  1          1   both   
12                    21.0       12427                  2          1   both   
13                    21.0       10258                  3          1   both   
14                    21.0       25133                  4          0   both   

    Unnamed: 0                             product_name  aisle_id  \
0        195.0                                     Soda      77.0   
1      14084.0  Organic Unsweetened Vanilla Almond Milk      91.0   
2      12427.0                      Original Beef Jerky      23.0   
3      26089.0               Aged White Cheddar Popcorn      23.0   
4      26406.0         XL Pick-A-Size Paper Towel Rolls      54.0   
5        195.0                                     Soda      77.0   
6      10258.0                               Pistachios     117.0   
7      12427.0                      Original Beef Jerky      23.0   
8      13176.0                   Bag of Organic Bananas      24.0   
9      26089.0               Aged White Cheddar Popcorn      23.0   
10     13032.0                    Cinnamon Toast Crunch     121.0   
11       195.0                                     Soda      77.0   
12     12427.0                      Original Beef Jerky      23.0   
13     10258.0                               Pistachios     117.0   
14     25134.0                    Organic String Cheese      21.0   

    department_id  prices     busiest_day busiest_period_of_day  max_order  
0             7.0     9.0  Regularly busy        Average orders         10  
1            16.0    12.5  Regularly busy        Average orders         10  
2            19.0     4.4  Regularly busy        Average orders         10  
3            19.0     4.7  Regularly busy        Average orders         10  


In [37]:
# Creating a 'Loyalty' Label Flag using loc Function
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [39]:
# Examining Value Counts on 'loyalty_flag'
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

Regular customer    15891507
Loyal customer      10294027
New customer         6249525
Name: loyalty_flag, dtype: int64

## Step 4. Analyze Spending Habits by Loyalty Category


In [40]:
# Determining Basic Statistics of Spending Habits by Loyalty Category
ords_prods_merge.groupby('loyalty_flag').agg({'prices' : ['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

## Comments:
The analysis of spending habits by loyalty category reveals that **loyal customers** spend the least on average ($10.39), followedby **regular customers** ($12.50) and **new customers** ($13.29). However, all categories have the same price range, indicating similar access to high-priced items across groups. This suggests opportunities to incentivize higher spending among loyal customers.

## Step 5. Create a Spending Flag

In [48]:
# Aggregating Data with transform() Function
ords_prods_merge['avg_product_price'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

In [49]:
# Checking Relevant Columns 
ords_prods_merge[['user_id', 'avg_product_price']].head(15)

user_id  avg_product_price
0         1           6.367797
1         1           6.367797
2         1           6.367797
3         1           6.367797
4         1           6.367797
5         1           6.367797
6         1           6.367797
7         1           6.367797
8         1           6.367797
9         1           6.367797
10        1           6.367797
11        1           6.367797
12        1           6.367797
13        1           6.367797
14        1           6.367797

In [51]:
# Creating a 'Spending' Label Flag using loc Function
ords_prods_merge.loc[ords_prods_merge['avg_product_price'] >= 10, 'spending_flag'] = 'High spender'
ords_prods_merge.loc[ords_prods_merge['avg_product_price'] < 10, 'spending_flag'] = 'Low spender'

In [52]:
# Examining Value Count on 'spending_flag' Column
ords_prods_merge['spending_flag'].value_counts(dropna = False)

Low spender     31800381
High spender      634678
Name: spending_flag, dtype: int64

In [53]:
# Checking Releavent Columns
ords_prods_merge[['user_id', 'avg_product_price', 'spending_flag']].head(15)

user_id  avg_product_price spending_flag
0         1           6.367797   Low spender
1         1           6.367797   Low spender
2         1           6.367797   Low spender
3         1           6.367797   Low spender
4         1           6.367797   Low spender
5         1           6.367797   Low spender
6         1           6.367797   Low spender
7         1           6.367797   Low spender
8         1           6.367797   Low spender
9         1           6.367797   Low spender
10        1           6.367797   Low spender
11        1           6.367797   Low spender
12        1           6.367797   Low spender
13        1           6.367797   Low spender
14        1           6.367797   Low spender

## Comments:
The data shows that **user_id 1** consistently has an average product price of **$6.37** and is categorized as a **Low spender**. This indicates uniform spending behavior for this user, suggesting stable purchasing habits within this category.

## Step 6. Create an Order Frequency Flag


In [54]:
# Aggregating Data with transform() Function
ords_prods_merge['median_days_between_orders'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

In [55]:
# Checking Relevant Columns 
ords_prods_merge[['user_id', 'days_since_prior_order', 'median_days_between_orders']].head(15)

user_id  days_since_prior_order  median_days_between_orders
0         1                     NaN                        20.5
1         1                     NaN                        20.5
2         1                     NaN                        20.5
3         1                     NaN                        20.5
4         1                     NaN                        20.5
5         1                    15.0                        20.5
6         1                    15.0                        20.5
7         1                    15.0                        20.5
8         1                    15.0                        20.5
9         1                    15.0                        20.5
10        1                    15.0                        20.5
11        1                    21.0                        20.5
12        1                    21.0                        20.5
13        1                    21.0                        20.5
14        1                    21.0                        20.5

In [58]:
# Creating a 'Order Frequency' Label Flag using loc Function
ords_prods_merge.loc[ords_prods_merge['median_days_between_orders'] > 20, 'order_frequency_flag'] = 'Non-frequent customer'
ords_prods_merge.loc[(ords_prods_merge['median_days_between_orders'] <= 20) & (ords_prods_merge['median_days_between_orders'] > 10), 'order_frequency_flag'] = 'Regular customer'
ords_prods_merge.loc[ords_prods_merge['median_days_between_orders'] <= 10, 'order_frequency_flag'] = 'Frequent customer'

In [59]:
# Examining Value Counts on 'order_frequency_flag'
ords_prods_merge['order_frequency_flag'].value_counts(dropna = False)

Frequent customer        21578273
Regular customer          7217066
Non-frequent customer     3639720
Name: order_frequency_flag, dtype: int64

In [60]:
# Checking Relevant Columns of Updated Dataframe
ords_prods_merge[['user_id', 'days_since_prior_order', 'median_days_between_orders', 'order_frequency_flag']].head(15)

user_id  days_since_prior_order  median_days_between_orders  \
0         1                     NaN                        20.5   
1         1                     NaN                        20.5   
2         1                     NaN                        20.5   
3         1                     NaN                        20.5   
4         1                     NaN                        20.5   
5         1                    15.0                        20.5   
6         1                    15.0                        20.5   
7         1                    15.0                        20.5   
8         1                    15.0                        20.5   
9         1                    15.0                        20.5   
10        1                    15.0                        20.5   
11        1                    21.0                        20.5   
12        1                    21.0                        20.5   
13        1                    21.0                        20.5   
14        1                    21.0                        20.5   

     order_frequency_flag  
0   Non-frequent customer  
1   Non-frequent customer  
2   Non-frequent customer  
3   Non-frequent customer  
4   Non-frequent customer  
5   Non-frequent customer  
6   Non-frequent customer  
7   Non-frequent customer  
8   Non-frequent customer  
9   Non-frequent customer  
10  Non-frequent customer  
11  Non-frequent customer  
12  Non-frequent customer  
13  Non-frequent customer  
14  Non-frequent customer

## Comments: 
The data shows that **user_id 1** has a median of **20.5 days between orders** and is categorized as a **Non-frequent customer**. Variations in `days_since_prior_order` (e.g., 15.0 and 21.0 days) align with the non-frequent shopping pattern. This indicates irregular but predictable ordering habits.

## Step 7. Export the Dataframe in pickle Format to the "Prepared Data" Folder

In [61]:
# Exporting the Dataframe
ords_prods_merge.to_pickle(os.path.join(path, 'Prepared Data', 'ords_prods_merge_grouped.pkl'))